### PR TITLE
[WIP] Support CPU fallback for unsupported ops in dynamo

### PR DESCRIPTION
**Pull request by wonjoolee95**
Fixes https://github.com/pytorch/xla/issues/4742

---

[WIP] Support CPU fallback for unsupported ops in dynamo

TODO
- [ ] Verify correctness and add unit tests
- [ ] Support shape/input based fallback
- [ ] Add more comments

### CODE REVIEW

**Comment by wonjoolee95** [comment]
@seanlatias, a quick verification that we can start with would be check if dynamo works with some of the unsupported ops. For example, something like:
```
import torch
import torch_xla

import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met
import torch._dynamo as dynamo

@dynamo.optimize("torchxla_trace_once")
def fn_fallback(M, mat1, mat2, beta):
  # xla currently only support alpha and beta == 1
  return torch.addmm(M, mat1, mat2, beta=beta)


M = torch.randn(2, 3, device=xm.xla_device())
mat1 = torch.randn(2, 3, device=xm.xla_device())
mat2 = torch.randn(3, 3, device=xm.xla_device())

res = fn_fallback(M, mat1, mat2, 0.5)
```

One more thing to verify would be to check if the `is_node_supported` function working as intended. We need to double check if the lines:
```
    return node.op == "call_function" and (getattr(node.target, "xla",
                                                   None) is not None or
                                           node.target == operator.getitem)
```
are enough to check if the node is supported by xla.

**Comment by seanlatias** [comment]
@wonjoolee95 is there a known list of unsupported ops? The example you show doesn't work. I think it is because of the limitation you mentioned before: the current implementation could not check the operator-operand combination.

**Comment by wonjoolee95** [comment]
Ah that's right, good catch. That's still a todo for this PR. So all ops that are supported by PyTorch/XLA are listed at https://github.com/pytorch/xla/blob/master/xla_native_functions.yaml. And all PyTorch ops are listed at https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml. So any op that's not in the PyTorch/XLA yaml file is not supported.

**Comment by seanlatias** [comment]
With the following code, I identified two issues.
```python
import torch
import torch_xla
    
import torch_xla.core.xla_model as xm
import torch_xla.debug.metrics as met
import torch._dynamo as dynamo

@dynamo.optimize("torchxla_trace_once")
def fn_fallback(M):
    return torch.cummin(M, 1)                      
                                           
M = torch.randn(5, 10, device=xm.xla_device())
res = fn_fallback(M)                                       
```

1. We need to enable the `allows_single_node_partition`. Otherwise there would be no partition here.
2. The `partitioner.fuse_partitions()` seems not working properly. The fused graph is simply wrong. I'll see what's the problem.
3. Shouldn't we move the fallback ops to CPU? Or we just compile the ops that can be offloaded of XLA. In that case we should fuse the graph at a later stage. If we just partition and fuse, we will result in the same graph even if the `fuse_partitions()` function works properly.

**Comment by seanlatias** [comment]
I think I find a workaround. Will push the code soon.

**Comment by wonjoolee95** [comment]
Thanks for the investigation, @seanlatias. I guess another simple test case would have the `torchxla_trace_once` wrapped under a function with a supported op *and* an unsupported op. 

Also, regarding the `allows_single_node_partition`, the `CapacityBasedPartitioner` seems to accept a parameter that controls this -- https://github.com/pytorch/pytorch/blob/master/torch/fx/passes/infra/partitioner.py#L39. And it is `False` by default, so we many manually want to turn that flag to `True`.

File: torch_xla/core/dynamo_bridge.py [hunk 1]
```diff
@@ -217,6 +241,9 @@ def extract_compiled_graph(xla_model: torch.fx.GraphModule, xla_args):
       tensor_id: i for i, tensor_id in enumerate(args_tensor_ids)
   }
 
+  # partition the graph for unsupported ops
+  xla_model = maybe_partition_graph(xla_model)
```
Comment (JackCaoG) [inline torch_xla/core/dynamo_bridge.py:245]: now `xla_model` becomes a `partitioned_graph`, and what happen next? In below code we will trace through this model `xla_out = xla_model(*xla_args)`, are we going to get a single graph and a single hash?

**Pull request closed without merging**